File: Algorithm/BFSAndDFS/BFS01.scala
package BFSAndDFS

import scala.collection.mutable

object BFS01 {
    def bfs(graph: Array[Array[Int]], start: Int, visited: Array[Boolean]): Unit = {
        val queue = mutable.Queue[Int]()
        queue.enqueue(start)
        visited(start) = true

        while (queue.nonEmpty) {
            val v = queue.dequeue()
            printf("%d ", v)
            for (i <- graph(v)) {
                if (!visited(i)) {
                    queue.enqueue(i)
                    visited(i) = true
                }
            }
        }
    }

    def main(args: Array[String]): Unit = {
        val graph = Array(
            Array(0),
            Array(2, 3, 8),
            Array(1, 7),
            Array(1, 4, 5),
            Array(3, 5),
            Array(3, 4),
            Array(7),
            Array(2, 6, 8),
            Array(1, 7)
        )
        var visited = Array.ofDim[Boolean](9)

        bfs(graph, 1, visited)
        println()
    }
}

File: Algorithm/BFSAndDFS/BFS01.go
package main

import "fmt"

type Queue []int

var (
	graph = [9][]int{
		{},
		{2, 3, 8},
		{1, 7},
		{1, 4, 5},
		{3, 5},
		{3, 4},
		{7},
		{2, 6, 8},
		{1, 7},
	}
	visited = [9]bool{}
)

func (q *Queue) enQueue(elem int) {
	*q = append(*q, elem)
}

func (q *Queue) deQueue() int {
	elem := (*q)[0]
	*q = (*q)[1:]
	return elem
}

func (q *Queue) nonEmpty() bool {
	return len(*q) != 0
}

func bfs(queue Queue, start int, visited []bool) {
	queue.enQueue(start)
	visited[start] = true

	for queue.nonEmpty() {
		v := queue.deQueue()
		fmt.Printf("%d ", v)
		for _, i := range graph[v] {
			if !visited[i] {
				queue.enQueue(i)
				visited[i] = true
			}
		}
	}
}

func main() {
	var queue Queue

	bfs(queue, 1, visited[:])
	fmt.Println()
}
